ssl/statem/statem_clnt.c

```diff
@@ -191,7 +191,7 @@ static int ossl_statem_client13_read_transition(SSL_CONNECTION *s, int mt)
         if (mt == SSL3_MT_CERTIFICATE_REQUEST) {
 #if DTLS_MAX_VERSION_INTERNAL != DTLS1_2_VERSION
             /* Restore digest for PHA before adding message.*/
-# error Internal DTLS version error
+# warning Internal DTLS version error
 #endif
```

Comment:
Shouldn't this if/endif just be deleted?

Reply:
I actually think this TODO is now obsolete.

It was originally added in 9d75dce3e. At that time the read state machine code looked like this:

https://github.com/openssl/openssl/blob/9d75dce3e1f94be6006500089491cb3284f70d06/ssl/statem/statem.c#L565-L624

Note in this snippet that near the top, for TLS, we get the message header. In the middle we transition the state machine state and then near the end we read the TLS message body. However for DTLS we just get the whole message near the top, and then transition after.

Importantly in `tls_get_message_body()` we call `ssl3_finish_mac` to feed the message into the handshake digest - and this happens *after* the state machine has transitioned. However is DTLS it all happens in `dtls_get_message` which happens before the transition. This is what the comment in this function a little below this is talking about.

Fast forward to today and the code looks different. This is the equivalent code in the read state machine:

https://github.com/openssl/openssl/blob/40a24c20a809916b43116c2bb16a36bdc40221f3/ssl/statem/statem.c#L616-L673

In this code we still get the DTLS message all in one go at the top, but there is now a new function after the transition called `dtls_get_message_body`...it doesn't actually get the body but it does do post-processing of the body including feeding the message into the handshake hash, so that the message is hashed *after* the transition in the same way as for TLS.

In other words DTLS and TLS now work the same way in this regards and therefore we don't need to do anything special for DTLSv1.3 here.

Reply:
So if I understand you correctly this
`
#if DTLS_MAX_VERSION_INTERNAL != DTLS1_2_VERSION
            /* Restore digest for PHA before adding message.*/
# warning Internal DTLS version error
#endif
            if (!SSL_CONNECTION_IS_DTLS(s)
                && s->post_handshake_auth == SSL_PHA_EXT_SENT) {
                s->post_handshake_auth = SSL_PHA_REQUESTED;
                /*
                 * In TLS, this is called before the message is added to the
                 * digest. In DTLS, this is expected to be called after adding
                 * to the digest. Either move the digest restore, or add the
                 * message here after the swap, or do it after the clientFinished?
                 */
                if (!tls13_restore_handshake_digest_for_pha(s)) {
                    /* SSLfatal() already called */
                    return 0;
                }
                st->hand_state = TLS_ST_CR_CERT_REQ;
                return 1;
            }
`
Should be:
`
            if (s->post_handshake_auth == SSL_PHA_EXT_SENT) {
                s->post_handshake_auth = SSL_PHA_REQUESTED;
                if (!tls13_restore_handshake_digest_for_pha(s)) {
                    /* SSLfatal() already called */
                    return 0;
                }
                st->hand_state = TLS_ST_CR_CERT_REQ;
                return 1;
            }
`
?

Reply:
Yes, I believe so.